src/rust/src/lib.rs

```diff
@@ -1,132 +1,224 @@
-    let mut col_offset = 0;
-    for col in processed_columns {
-        let n = col.ncols();
-        result.slice_mut(s![.., col_offset..col_offset+n]).assign(&col);
-        col_offset += n;
-    }
-    let nrows = result.nrows();
-    let mut robj: Robj = result.try_into().unwrap();
+    let mut robj: Robj = processed_columns_matrix.into();
 
     // Create dimnames list
-    let row_names: Vec<String> = (1..=nrows).map(|i| i.to_string()).collect();
-    let dimnames = List::from_values(&[row_names, column_names.clone()]);
+    let row_names: Vec<String> = (1..=nrow).map(|i| i.to_string()).collect();
+    let dimnames = List::from_values(&[row_names, column_names]);
 
     // Set dimnames attribute
     robj.set_attrib("dimnames", dimnames)?;
 
     Ok(robj)
 }
 
-fn process_integer_column(column: &Robj, col_name: &str, nrow: usize) -> (Array2<f64>, Vec<String>) {
-    if column.inherits("factor") {
-        process_factor_column(column, col_name, nrow)
-    } else {
-        let int_col: Vec<i32> = column.as_integer_vector().unwrap();
-        let float_col: Array2<f64> = Array2::from_shape_vec((nrow, 1), int_col.into_iter().map(|x| x as f64).collect()).unwrap();
-        (float_col, vec![col_name.to_string()])
-    }
+fn process_integer_column(
+    column: &Robj,
+    col_name: &str,
+    output_column_names: &mut Vec<String>,
+    output: &mut [f64],
+) {
+    output_column_names.push(col_name.to_string());
+    zip(column.as_integer_slice().unwrap().iter(), output.iter_mut()).for_each(
+        |(integer_element, output)| {
+            *output = *integer_element as f64;
+        },
+    );
 }
 
-fn process_factor_column(column: &Robj, col_name: &str, nrow: usize) -> (Array2<f64>, Vec<String>) {
-    let int_col: Vec<i32> = column.as_integer_vector().unwrap();
-    let levels: Vec<String> = column.levels().unwrap().map(|s| s.to_string()).collect();
-    let mut dummy_cols = Array2::<f64>::zeros((nrow, levels.len() - 1));
-
-    for (i, &val) in int_col.iter().enumerate() {
-        if val > 1 && val <= levels.len() as i32 {
-            let level_index = (val - 2) as usize;
-            if level_index < dummy_cols.ncols() {
-                dummy_cols[[i, level_index]] = 1.0;
-            }
+fn process_factor_column(
+    column: &Robj,
+    col_name: &str,
+    nrow: usize,
+    output_column_names: &mut Vec<String>,
+    output: &mut [f64],
+) {
+    output_column_names.extend(
+        column
+            .levels()
+            .unwrap()
+            .skip(1)
+            .map(|level| format!("{}_{}", col_name, level)),
+    );
+    // remove the first level from all the factors
+    let level_index = column.as_integer_slice().unwrap().iter().map(|x| *x - 2);
+    // let nlevels = column.levels().unwrap().len() - 1;
+    output.fill(0.);
+    for (k, level_index) in level_index.enumerate() {
+        let row_id = k;
+        // dbg!(k, level_index);
+        let col_id: Option<usize> = level_index.try_into().ok();
```

Comment:
Since we can have a negative value, which needs to be ignored, I use the conversion as a way to ignore it.